addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js

```diff
@@ -3980,6 +3994,89 @@ export class OdooEditor extends EventTarget {
+        }
+        while (nodeBeforeCursor && nodesToSkip.includes(nodeBeforeCursor.nodeName)) {
+            nodeBeforeCursor = nodeBeforeCursor.previousSibling;
+        }
+        // Utils
+        const restoreLastSelection = () => {
+            const lastSelection = this._latestComputedSelectionInEditable;
+            if (!lastSelection || lastSelection.anchorNode === this.editable) {
+                selection.removeAllRanges();
+                return;
+            }
+            selection.setBaseAndExtent(lastSelection.anchorNode, lastSelection.anchorOffset,
+                lastSelection.focusNode, lastSelection.focusOffset);
+        }
+        const insertParagraph = (element, method) => {
+            this._recordHistorySelection(true);
+            // Prevent saving a history step with selection at root.
+            if (this._currentStep.selection.anchorNodeOid === 'root') {
+                this._currentStep.selection = {};
+            }
+            const p = this.document.createElement('p');
+            p.append(this.document.createElement('br'));
+            element[method](p);
+            this.historyStep();
+            if (this._currentStep.selection.anchorNodeOid === 'root') {
+                this._currentStep.selection = {};
+            }
+            return p;
+        }
+        // Handle arrow key presses. Do not insert new 'P'.
+        if (currentKeyPress === 'ArrowRight' || currentKeyPress === 'ArrowDown') {
+            if (nodeAfterCursor) {
+                setSelection(...getDeepestPosition(nodeAfterCursor, 0));
+            } else {
+                restoreLastSelection();
+            }
+        } else if (currentKeyPress === 'ArrowLeft' || currentKeyPress === 'ArrowUp') {
+            if (nodeBeforeCursor) {
+                setSelection(...getDeepestPosition(nodeBeforeCursor, nodeSize(nodeBeforeCursor)));
+            } else {
+                restoreLastSelection();
+            }
+        // Handle other forms of selection. Insert a 'P' if necessary.
+        } else if (!nodeAfterCursor) {
+            // Cursor is at the end of the editable.
+            const p = insertParagraph(this.editable, 'append');
+            setCursorStart(p);
+        } else if (!nodeBeforeCursor) {
+            // Cursor is at the beginning of the editable.
+            const p = insertParagraph(this.editable, 'prepend');
+            setCursorStart(p);
+        } else if (paragraphRelatedElements.includes(nodeAfterCursor.nodeName)) {
+            // Cursor is right before a p.
+            setCursorStart(nodeAfterCursor);
+        } else if (paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
+            // Cursor is right after a p.
+            setCursorEnd(nodeBeforeCursor);
+        } else {
+            // Cursor is between two non-p blocks
+            const p = insertParagraph(nodeAfterCursor, 'before');
+            setCursorStart(p);
+        }
+    }
```

Comment:
I suggest the following rewrite to get rid of the local function:
```js
if (nodeAfterCursor && paragraphRelatedElements.includes(nodeAfterCursor.nodeName)) {
    // Cursor is right before a p.
    setCursorStart(nodeAfterCursor);
} else if (nodeBeforeCursor && paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
    // Cursor is right after a p.
    setCursorEnd(nodeBeforeCursor);
} else {
    this._recordHistorySelection(true);
    // Prevent saving a history step with selection at root.
    if (this._currentStep.selection.anchorNodeOid === 'root') {
        this._currentStep.selection = {};
    }
    const p = this.document.createElement('p');
    p.append(this.document.createElement('br'));
    if (!nodeAfterCursor) {
        // Cursor is at the end of the editable.
        element.append(p);
    } else if (!nodeBeforeCursor) {
        // Cursor is at the beginning of the editable.
        element.prepend(p);
    } else {
        // Cursor is between two non-p blocks
        element.before(p);
    }
    element[method](p);
    this.historyStep();
    if (this._currentStep.selection.anchorNodeOid === 'root') {
        this._currentStep.selection = {};
    }
    setCursorStart(p);
}
```

Reply:
Good one!